rbd: fix vol.VolID in cloneFromSnapshot()

Rakshith-R opened the pull request:
Volume generated from snap using genrateVolFromSnap
already copies volume ID correctly, therefore removing
`vol.VolID = rbdVol.VolID` which wrongly copies parent
Volume ID instead leading to error from copyEncryption()
on parent and clone volume ID being equal.

Signed-off-by: Rakshith R <rar@redhat.com>

<details>
<summary>
Logs
</summary>

```
I0709 05:48:19.899510       1 identityserver-default.go:38] ID: 20 Using default GetPluginInfo
I0709 05:48:19.899585       1 utils.go:182] ID: 20 GRPC response: {"name":"rook-ceph.rbd.csi.ceph.com","vendor_version":"canary"}
I0709 05:48:19.903851       1 utils.go:171] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC call: /csi.v1.Controller/CreateSnapshot
I0709 05:48:19.904174       1 utils.go:175] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC request: {"name":"snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2","parameters":{"clusterID":"rook-ceph"},"secrets":"***stripped***","source_volume_id":"0001-0009-rook-ceph-0000000000000001-9c4c4997-e078-11eb-9aa0-0242ac110005"}
I0709 05:48:19.918382       1 omap.go:84] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 got omap values: (pool="replicapool", namespace="", name="csi.volume.9c4c4997-e078-11eb-9aa0-0242ac110005"): map[csi.imageid:116b87d2743c csi.imagename:csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005 csi.volname:pvc-4dc3aa8b-1fea-4e78-9f85-9beecf39639a csi.volume.encryptKMS:user-ns-secrets-metadata-test csi.volume.owner:rook-ceph]
E0709 05:48:19.937789       1 omap.go:77] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 omap not found (pool="replicapool", namespace="", name="csi.snaps.default"): rados: ret=-2, No such file or directory
I0709 05:48:19.952186       1 omap.go:148] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 set omap keys (pool="replicapool", namespace="", name="csi.snaps.default"): map[csi.snap.snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2:43e1f4b4-e079-11eb-bda3-0242ac110005])
I0709 05:48:19.954816       1 omap.go:148] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 set omap keys (pool="replicapool", namespace="", name="csi.snap.43e1f4b4-e079-11eb-bda3-0242ac110005"): map[csi.imagename:csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 csi.snapname:snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 csi.source:csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005 csi.volume.encryptKMS:user-ns-secrets-metadata-test csi.volume.owner:rook-ceph])
I0709 05:48:19.954834       1 rbd_journal.go:408] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 generated Volume ID (0001-0009-rook-ceph-0000000000000001-43e1f4b4-e079-11eb-bda3-0242ac110005) and image name (csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005) for request name (snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2)
I0709 05:48:19.954868       1 rbd_util.go:1142] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 rbd: snap create replicapool/csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005@csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 using mon 192.168.39.136:6789
I0709 05:48:20.162401       1 rbd_util.go:1199] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 rbd: clone replicapool/csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005@csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 replicapool/csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 (features: [layering deep-flatten]) using mon 192.168.39.136:6789
I0709 05:48:20.201395       1 rbd_util.go:1154] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 rbd: snap rm replicapool/csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005@csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 using mon 192.168.39.136:6789
I0709 05:48:20.423415       1 encryption.go:75] image replicapool/csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005 encrypted state metadata reports "encryptionPrepared"
I0709 05:48:20.498974       1 omap.go:148] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 set omap keys (pool="replicapool", namespace="", name="csi.snap.43e1f4b4-e079-11eb-bda3-0242ac110005"): map[csi.imageid:11c953ba8286])
I0709 05:48:20.529096       1 rbd_util.go:677] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 clone depth is (1), configured softlimit (4) and hardlimit (8) for replicapool/csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005
I0709 05:48:20.529362       1 utils.go:182] ID: 21 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC response: {"snapshot":{"creation_time":{"nanos":176795825,"seconds":1625809700},"size_bytes":1073741824,"snapshot_id":"0001-0009-rook-ceph-0000000000000001-43e1f4b4-e079-11eb-bda3-0242ac110005","source_volume_id":"0001-0009-rook-ceph-0000000000000001-9c4c4997-e078-11eb-9aa0-0242ac110005"}}
I0709 05:48:20.568957       1 utils.go:171] ID: 22 GRPC call: /csi.v1.Identity/GetPluginInfo
I0709 05:48:20.568997       1 utils.go:175] ID: 22 GRPC request: {}
I0709 05:48:20.569020       1 identityserver-default.go:38] ID: 22 Using default GetPluginInfo
I0709 05:48:20.569040       1 utils.go:182] ID: 22 GRPC response: {"name":"rook-ceph.rbd.csi.ceph.com","vendor_version":"canary"}
I0709 05:48:20.569539       1 utils.go:171] ID: 23 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC call: /csi.v1.Controller/CreateSnapshot
I0709 05:48:20.569608       1 utils.go:175] ID: 23 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC request: {"name":"snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2","parameters":{"clusterID":"rook-ceph"},"secrets":"***stripped***","source_volume_id":"0001-0009-rook-ceph-0000000000000001-9c4c4997-e078-11eb-9aa0-0242ac110005"}
I0709 05:48:20.570437       1 omap.go:84] ID: 23 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 got omap values: (pool="replicapool", namespace="", name="csi.volume.9c4c4997-e078-11eb-9aa0-0242ac110005"): map[csi.imageid:116b87d2743c csi.imagename:csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005 csi.volname:pvc-4dc3aa8b-1fea-4e78-9f85-9beecf39639a csi.volume.encryptKMS:user-ns-secrets-metadata-test csi.volume.owner:rook-ceph]
I0709 05:48:20.596760       1 omap.go:84] ID: 23 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 got omap values: (pool="replicapool", namespace="", name="csi.snaps.default"): map[csi.snap.snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2:43e1f4b4-e079-11eb-bda3-0242ac110005]
I0709 05:48:20.597293       1 omap.go:84] ID: 23 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 got omap values: (pool="replicapool", namespace="", name="csi.snap.43e1f4b4-e079-11eb-bda3-0242ac110005"): map[csi.imageid:11c953ba8286 csi.imagename:csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 csi.snapname:snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 csi.source:csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005 csi.volume.encryptKMS:user-ns-secrets-metadata-test csi.volume.owner:rook-ceph]
I0709 05:48:20.639290       1 rbd_util.go:1142] ID: 23 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 rbd: snap create replicapool/csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005@csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 using mon 192.168.39.136:6789
E0709 05:48:21.169785       1 utils.go:180] ID: 23 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC error: rpc error: code = Internal desc = snapshot not found: snap replicapool/csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005@csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 not found
I0709 05:48:21.179357       1 utils.go:171] ID: 24 GRPC call: /csi.v1.Identity/GetPluginInfo
I0709 05:48:21.179534       1 utils.go:175] ID: 24 GRPC request: {}
I0709 05:48:21.179545       1 identityserver-default.go:38] ID: 24 Using default GetPluginInfo
I0709 05:48:21.179605       1 utils.go:182] ID: 24 GRPC response: {"name":"rook-ceph.rbd.csi.ceph.com","vendor_version":"canary"}
I0709 05:48:21.182207       1 utils.go:171] ID: 25 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC call: /csi.v1.Controller/CreateSnapshot
I0709 05:48:21.182402       1 utils.go:175] ID: 25 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC request: {"name":"snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2","parameters":{"clusterID":"rook-ceph"},"secrets":"***stripped***","source_volume_id":"0001-0009-rook-ceph-0000000000000001-9c4c4997-e078-11eb-9aa0-0242ac110005"}
I0709 05:48:21.189749       1 omap.go:84] ID: 25 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 got omap values: (pool="replicapool", namespace="", name="csi.volume.9c4c4997-e078-11eb-9aa0-0242ac110005"): map[csi.imageid:116b87d2743c csi.imagename:csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005 csi.volname:pvc-4dc3aa8b-1fea-4e78-9f85-9beecf39639a csi.volume.encryptKMS:user-ns-secrets-metadata-test csi.volume.owner:rook-ceph]
I0709 05:48:21.227420       1 omap.go:84] ID: 25 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 got omap values: (pool="replicapool", namespace="", name="csi.snaps.default"): map[csi.snap.snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2:43e1f4b4-e079-11eb-bda3-0242ac110005]
I0709 05:48:21.227920       1 omap.go:84] ID: 25 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 got omap values: (pool="replicapool", namespace="", name="csi.snap.43e1f4b4-e079-11eb-bda3-0242ac110005"): map[csi.imageid:11c953ba8286 csi.imagename:csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005 csi.snapname:snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 csi.source:csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005 csi.volume.encryptKMS:user-ns-secrets-metadata-test csi.volume.owner:rook-ceph]
I0709 05:48:21.256230       1 rbd_journal.go:233] ID: 25 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 found existing image (0001-0009-rook-ceph-0000000000000001-43e1f4b4-e079-11eb-bda3-0242ac110005) with name (csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005) for request (snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2)
E0709 05:48:21.256484       1 utils.go:180] ID: 25 Req-ID: snapshot-bba33d6c-d407-4bfa-a6c9-e6238e6750c2 GRPC error: rpc error: code = Internal desc = BUG: "replicapool/csi-vol-9c4c4997-e078-11eb-9aa0-0242ac110005" and "replicapool/csi-snap-43e1f4b4-e079-11eb-bda3-0242ac110005" have the same VolID (0001-0009-rook-ceph-0000000000000001-9c4c4997-e078-11eb-9aa0-0242ac110005) set!? Call stack: goroutine 93 [running]:
github.com/ceph/ceph-csi/internal/util.CallStack(...)
	/go/src/github.com/ceph/ceph-csi/internal/util/util.go:349
github.com/ceph/ceph-csi/internal/rbd.(*rbdImage).copyEncryptionConfig(0xc000620540, 0xc0006208c0, 0x0, 0x0)
	/go/src/github.com/ceph/ceph-csi/internal/rbd/encryption.go:121 +0x551
github.com/ceph/ceph-csi/internal/rbd.cloneFromSnapshot(0x1fb9658, 0xc0006cf0b0, 0xc000620540, 0xc000666240, 0xc0004b0360, 0x0, 0x0, 0x0)
	/go/src/github.com/ceph/ceph-csi/internal/rbd/controllerserver.go:1001 +0x219
github.com/ceph/ceph-csi/internal/rbd.(*ControllerServer).CreateSnapshot(0xc00031e7a0, 0x1fb9658, 0xc0006cf0b0, 0xc000676af0, 0x0, 0x0, 0x0)
	/go/src/github.com/ceph/ceph-csi/internal/rbd/controllerserver.go:935 +0xc8c
github.com/container-storage-interface/spec/lib/go/csi._Controller_CreateSnapshot_Handler.func1(0x1fb9658, 0xc0006cf0b0, 0x1bb6cc0, 0xc000676af0, 0x0, 0x0, 0x1, 0xc0003a9040)
	/go/src/github.com/ceph/ceph-csi/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:5820 +0x89
github.com/ceph/ceph-csi/internal/csi-common.panicHandler(0x1fb9658, 0xc0006cf0b0, 0x1bb6cc0, 0xc000676af0, 0xc0004b02a0, 0xc00000e7c8, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:199 +0x96
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1fb9658, 0xc0006cf0b0, 0x1bb6cc0, 0xc000676af0, 0xc0000e3980, 0x1, 0x1, 0x30)
	/go/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25 +0x63
github.com/ceph/ceph-csi/internal/csi-common.logGRPC(0x1fb9658, 0xc0006cf0b0, 0x1bb6cc0, 0xc000676af0, 0xc0004b02a0, 0xc0004b0300, 0x1b7e8c0, 0xc0006cf0b0, 0x19e1f80, 0xc0003a8e40)
	/go/src/github.com/ceph/ceph-csi/internal/csi-common/utils.go:178 +0x206
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1fb9658, 0xc0006cf0b0, 0x1bb6cc0, 0xc000676af0, 0x19e8700, 0xc0003a8e50, 0x1fb9658, 0xc0006cf0b0)
	/go/src/github.com/ceph/ceph-csi/vendor/github.com/grpc-ecosystem/go-grpc-middleware/cha
```
</details>

Code change to simulate `readyToUse` being set to false (will happen when number of snapshots hits flatten limit)  https://github.com/Rakshith-R/ceph-csi/commit/2097d4fa87d7a91f9cb3ccad628cacb2b6fee384

Reproducible image: http://ghcr.io/rakshith-r/cephcsi:buggy-snap
Steps: 
- Create encrypted pvc
- Create encrypted snapshot

Fixes: #2285
Fixes: #1883 

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Rakshith-R committed:
rbd: fix vol.VolID in cloneFromSnapshot()

Volume generated from snap using genrateVolFromSnap
already copies volume ID correctly, therefore removing
`vol.VolID = rbdVol.VolID` which wrongly copies parent
Volume ID instead leading to error from copyEncryption()
on parent and clone volume ID being equal.

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented:
@Rakshith-R just a question if this can be easily reproduced why didn't get caught in CI? can you see anything we are missing in CI and anything need to be added?

humblec commented:
> @Rakshith-R just a question if this can be easily reproduced why didn't get caught in CI? can you see anything we are missing in CI and anything need to be added?

yeah, interesting that it was not caught on r our e2e... 
@Rakshith-R while you say `will happen when number of snapshots hits flatten limit` -> what is the limit you are mentioning ?

Rakshith-R commented:
> @Rakshith-R just a question if this can be easily reproduced why didn't get caught in CI? can you see anything we are missing in CI and anything need to be added?

Basically, most of our tests pass in one go in ci (a single request). 

This can only be caught when snapshotting process either returns a `readyTouse=false` (have to flatten) or process is killed in between(in some precise step of the control flow) basically leaving some omap entry/snaps/clones to continue from.

nixpanic reviewed (dismissed):
Indeed strange that this was not caught earlier. Maybe e2e testing is not complete?

Rakshith-R commented:
> > @Rakshith-R just a question if this can be easily reproduced why didn't get caught in CI? can you see anything we are missing in CI and anything need to be added?
> 
> yeah, interesting that it was not caught on r our e2e...
> @Rakshith-R while you say `will happen when number of snapshots hits flatten limit` -> what is the limit you are mentioning ?

Scratch ~~number of snapshots~~ , I meant what this function does https://github.com/ceph/ceph-csi/blob/b27d6319ca8b16004ecbc00f71393821594de712/internal/rbd/rbd_util.go#L663

Rakshith-R commented:
https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/rbd-snap-clone.md#rbd-cli-commands-to-create-snapshot
This step, which would trigger a follow up request from external-snapshotter
```
// check the depth, if the depth is greater than configured hardlimit add a
// task to flatten the cloned image, return snapshot status ready as `false`,
// if the depth is greater than softlimit add a task to flatten the image
// and return success
ceph rbd task add flatten <RBD image for temporary snap image>
```
Or can also be caused from a very well(or very ill) timed container restart.

Madhu-1 commented:
https://github.com/ceph/ceph-csi/blob/680a7bf411b58858d88f56d5a5a045baa2e743bb/scripts/install-helm.sh#L176

```
--set provisioner.maxSnapshotsOnImage=3 --set provisioner.minSnapshotsOnImage=2
```

I think we are missing to `rbdhardmaxclonedepth` and `rbdsoftmaxclonedepth` to 1 or 2 and see this can be caught in CI?

* Create PVC
* Create PVC clone (if required create chain of  clones)
* Create snapshot from clone -> this will cause the flatten to happen

can you see this helps?

Rakshith-R commented:
> https://github.com/ceph/ceph-csi/blob/680a7bf411b58858d88f56d5a5a045baa2e743bb/scripts/install-helm.sh#L176
> 
> ```
> --set provisioner.maxSnapshotsOnImage=3 --set provisioner.minSnapshotsOnImage=2
> ```
> 
> I think we are missing to `rbdhardmaxclonedepth` and `rbdsoftmaxclonedepth` to 1 or 2 and see this can be caught in CI?
> 
> * Create PVC
> * Create PVC clone (if required create chain of  clones)
> * Create snapshot from clone -> this will cause the flatten to happen
> 
> can you see this helps?

Or creating a long chain of snapshot /& clones. 
This ideally should trigger both soft and hard limit.


And as to snapshot limit, it is flattened at the very begining of request https://github.com/ceph/ceph-csi/blob/b27d6319ca8b16004ecbc00f71393821594de712/internal/rbd/controllerserver.go#L934 . I don't think it matters much. (referred to this instead of rbd image chain Depth by mistake :sweat_smile: )

Madhu-1 commented:
> > https://github.com/ceph/ceph-csi/blob/680a7bf411b58858d88f56d5a5a045baa2e743bb/scripts/install-helm.sh#L176
> > 
> > ```
> > --set provisioner.maxSnapshotsOnImage=3 --set provisioner.minSnapshotsOnImage=2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think we are missing to `rbdhardmaxclonedepth` and `rbdsoftmaxclonedepth` to 1 or 2 and see this can be caught in CI?
> > 
> > * Create PVC
> > * Create PVC clone (if required create chain of  clones)
> > * Create snapshot from clone -> this will cause the flatten to happen
> > 
> > can you see this helps?
> 
> Or creating a long chain of snapshot /& clones.
> This ideally should trigger both soft and hard limit.
> 
> And as to snapshot limit, it is flattened at the very begining of request
> 
> https://github.com/ceph/ceph-csi/blob/b27d6319ca8b16004ecbc00f71393821594de712/internal/rbd/controllerserver.go#L934
> 
> . I don't think it matters much. (referred to this instead of rbd image chain Depth by mistake 😅 )

Yes that also helps. Can you please add it

Madhu-1 commented:
Can you also add a commit to set hard and soft limit to lower values in helm? This will avoid creating larger chain to test this one

humblec commented:
> Or creating a long chain of snapshot /& clones.
> This ideally should trigger both soft and hard limit.
> 
> And as to snapshot limit, it is flattened at the very begining of request
> 
> https://github.com/ceph/ceph-csi/blob/b27d6319ca8b16004ecbc00f71393821594de712/internal/rbd/controllerserver.go#L934
> . I don't think it matters much. (referred to this instead of rbd image chain Depth by mistake sweat_smile )

Yeah, that clarifies it, :+1:

agarwal-mudit commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

agarwal-mudit commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

agarwal-mudit commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

obnoxxx commented:
> > @Rakshith-R just a question if this can be easily reproduced why didn't get caught in CI? can you see anything we are missing in CI and anything need to be added?
> 
> Basically, most of our tests pass in one go in ci (a single request).
> 
> This can only be caught when snapshotting process either returns a `readyTouse=false` (have to flatten) or process is killed in between(in some precise step of the control flow) basically leaving some omap entry/snaps/clones to continue from.

If I had anything to say here, I would request a fix like this to be accompanied by a commit adding or expanding a test case to prove that the pre-fix state is wrong, in this very PR! 😄 

More thoughts:

* If there is no regular/deterministic way of creating a snapshot resulting in `readyToUse = false`, then we should create this possibility in the test code.
* My gut feeling says it should be possible to cover this with a unit / subsystem test, should not require e2e test.

obnoxxx reviewed (changes requested):
I don't really have a voice here, but I would request to add a test case (as a separate commit in this same PR) for this if possible.

Rakshith-R commented:
> If I had anything to say here, I would request a fix like this to be accompanied by a commit adding or expanding a test case to prove that the pre-fix state is wrong, in this very PR! 

Modified args 
```
            - '--rbdhardmaxclonedepth=1'
            - '--rbdsoftmaxclonedepth=0'
```
Which will flatten all the clones/snaps/restores. 
> More thoughts:
> 
> * If there is no regular/deterministic way of creating a snapshot resulting in `readyToUse = false`, then we should create this possibility in the test code.

This will surely result in `readyToUse = false`, :crossed_fingers: hope e2e does not timeout though

> * My gut feeling says it should be possible to cover this with a unit / subsystem test, should not require e2e test.

Needs interaction with ceph cluster, so not possible.

Rakshith-R commented:
> I don't really have a voice here, but I would request to add a test case (as a separate commit in this same PR) for this if possible.

We also add a task to flatten image(during snap/clone/restore), so cannot effectively guarantee depth of image to be a certain value(at a certain time). 
But we can decrease the limit as I've done in helm_install.sh so it should act as e2e test.

obnoxxx commented:
> > I don't really have a voice here, but I would request to add a test case (as a separate commit in this same PR) for this if possible.
> 
> We also add a task to flatten image(during snap/clone/restore), so cannot effectively guarantee depth of image to be a certain value(at a certain time).
> But we can decrease the limit as I've done in helm_install.sh so it should act as e2e test.

But this is not an explicit test for this error, right? An explicit test would check the vo-ids after `cloneFromSnapshot` function. Are we sure this will fail e2e tests without the code fix?

Rakshith-R commented:
> But this is not an explicit test for this error, right? An explicit test would check the vo-ids after `cloneFromSnapshot` function. 
>Are we sure this will fail e2e tests without the code fix?

yes, it will fail. 
Tested with those arguments and here's the logs https://hackmd.io/7pDqtCupRtiZrRlrz_fHDw

Madhu-1 commented on scripts/install-helm.sh:
```diff
@@ -173,7 +173,7 @@ install_cephcsi_helm_charts() {
     kubectl delete cm ceph-csi-config --namespace ${NAMESPACE}
     # shellcheck disable=SC2086
-    "${HELM}" install --namespace ${NAMESPACE} --set provisioner.fullnameOverride=csi-rbdplugin-provisioner --set nodeplugin.fullnameOverride=csi-rbdplugin --set configMapName=ceph-csi-config --set provisioner.podSecurityPolicy.enabled=true --set nodeplugin.podSecurityPolicy.enabled=true --set provisioner.replicaCount=1 ${SET_SC_TEMPLATE_VALUES} ${RBD_SECRET_TEMPLATE_VALUES} ${RBD_CHART_NAME} "${ ...
+    "${HELM}" install --namespace ${NAMESPACE} --set provisioner.fullnameOverride=csi-rbdplugin-provisioner --set nodeplugin.fullnameOverride=csi-rbdplugin --set configMapName=ceph-csi-config --set provisioner.podSecurityPolicy.enabled=true --set nodeplugin.podSecurityPolicy.enabled=true --set provisioner.replicaCount=1 ${SET_SC_TEMPLATE_VALUES} ${RBD_SECRET_TEMPLATE_VALUES} ${RBD_CHART_NAME} "${ ...
```
set these values to 2 and 3 respectively. or else https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/nodeserver.go#L301-L306 will not get tested in nodeserver

Madhu-1 replied:
okay, discard this one. this will get tested in the non-helm scenario

Madhu-1 replied:
okay with this change a real bug was caught, looks like the Pvc-pvc clone is not working as expected and an image leak is happening.

Madhu-1 commented:
/test ci/centos/mini-e2e-helm/k8s-1.20

Madhu-1 commented:
@Rakshith-R can you please check PVC clones e2e.

humblec commented:
> > But this is not an explicit test for this error, right? An explicit test would check the vo-ids after `cloneFromSnapshot` function.
> > Are we sure this will fail e2e tests without the code fix?
> 
> yes, it will fail.
> Tested with those arguments and here's the logs https://hackmd.io/7pDqtCupRtiZrRlrz_fHDw

Tweaking the soft/hard  limits or both with allowed values would be one good to way to bring up the failure scenarios here. Looks like  we need to add few test cases here with different values which could help us to reveal the flow we are going wrong.

Madhu-1 committed:
rbd: fix clone problem

This commit fixes a bug in checkCloneImage() which was caused
by checking cloned image before checking on temp-clone image snap
in a subsequent request which lead to stale images. This was solved
by checking temp-clone image snap and flattening temp-clone if
needed.
This commit also fixes comparison bug in flattenCloneImage().

Signed-off-by: Rakshith R <rar@redhat.com>
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: copy creds when copying the connection

rbd flatten functions is a CLI call and it expects
the creds as the input and copying of creds is
required when we generate the temp clone image.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Rakshith-R committed:
cleanup: refractor checkCloneImage to reducing nesting if

This commit refractors checkCloneImage function to
address nestif linter issue.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add test cases for snapshot-restore chain with depth 2

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add test cases for pvc-pvcClone chain with depth 2

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
Both the following e2e is failing at `validate clones in different pool` e2e. 
[helm-1.21](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.21/detail/mini-e2e-helm_k8s-1.21/320/pipeline/ ) ,
[helm-1.20](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/1655/pipeline/)

Also https://github.com/ceph/ceph-csi/issues/1883 is hit at 

```
I0715 11:13:56.839829       1 utils.go:171] ID: 707 Req-ID: pvc-4e5cbe5f-1623-402e-982a-26a08b6da064 GRPC call: /csi.v1.Controller/CreateVolume
I0715 11:13:56.840227       1 utils.go:175] ID: 707 Req-ID: pvc-4e5cbe5f-1623-402e-982a-26a08b6da064 GRPC request: {"accessibility_requirements":{"preferred":[{"segments":{"topology.rbd.csi.ceph.com/region":"testregion","topology.rbd.csi.ceph.com/zone":"testzone"}}],"requisite":[{"segments":{"topology.rbd.csi.ceph.com/region":"testregion","topology.rbd.csi.ceph.com/zone":"testzone"}}]},"capacity_range":{"required_bytes":1073741824},"name":"pvc-4e5cbe5f-1623-402e-982a-26a08b6da064","parameters":{"clusterID":"cd8b81d5-27be-44a3-8230-d3e1a2369def","csi.storage.k8s.io/pv/name":"pvc-4e5cbe5f-1623-402e-982a-26a08b6da064","csi.storage.k8s.io/pvc/name":"rbd-32741","csi.storage.k8s.io/pvc/namespace":"rbd-3274","imageFeatures":"layering","pool":"clone-test","thickProvision":"false"},"secrets":"***stripped***","volume_capabilities":[{"AccessType":{"Mount":{"fs_type":"ext4","mount_flags":["discard"]}},"access_mode":{"mode":1}}],"volume_content_source":{"Type":{"Snapshot":{"snapshot_id":"0001-0024-cd8b81d5-27be-44a3-8230-d3e1a2369def-0000000000000002-60e3ec6e-e55d-11eb-9cdd-5ef793f1116d"}}}}
I0715 11:13:56.840502       1 rbd_util.go:1067] ID: 707 Req-ID: pvc-4e5cbe5f-1623-402e-982a-26a08b6da064 setting disableInUseChecks: false image features: [layering] mounter: rbd
E0715 11:13:56.841579       1 omap.go:77] ID: 707 Req-ID: pvc-4e5cbe5f-1623-402e-982a-26a08b6da064 omap not found (pool="replicapool", namespace="", name="csi.snap.60e3ec6e-e55d-11eb-9cdd-5ef793f1116d"): rados: ret=-2, No such file or directory
W0715 11:13:56.841620       1 voljournal.go:672] ID: 707 Req-ID: pvc-4e5cbe5f-1623-402e-982a-26a08b6da064 unable to read omap keys: pool or key missing: key not found: rados: ret=-2, No such file or directory
E0715 11:13:56.841643       1 controllerserver.go:682] ID: 707 Req-ID: pvc-4e5cbe5f-1623-402e-982a-26a08b6da064 failed to get backend snapshot for 0001-0024-cd8b81d5-27be-44a3-8230-d3e1a2369def-0000000000000002-60e3ec6e-e55d-11eb-9cdd-5ef793f1116d: key not found: no snap source in omap for "csi.snap.60e3ec6e-e55d-11eb-9cdd-5ef793f1116d"
E0715 11:13:56.841725       1 utils.go:180] ID: 707 Req-ID: pvc-4e5cbe5f-1623-402e-982a-26a08b6da064 GRPC error: rpc error: code = Internal desc = key not found: no snap source in omap for "csi.snap.60e3ec6e-e55d-11eb-9cdd-5ef793f1116d"
```

cc @Yuggupta27

Rakshith-R commented:
/retest all

Rakshith-R commented:
> /retest all

retesting to capture more such fails.

Rakshith-R committed:
e2e: log imageList in validateRBDImageCount for better debugging

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented:
@Rakshith-R PR need manual rebase

Rakshith-R committed:
rbd: fix snapshot id idempotency issue

This commit fixes snapshot id idempotency issue by
always returning an error when flattening is in progress
and not using `readyToUse:false` response.

Signed-off-by: Rakshith R <rar@redhat.com>

nixpanic commented on internal/rbd/clone.go:
```diff
@@ -52,31 +52,7 @@ func (rv *rbdVolume) checkCloneImage(ctx context.Context, parentVol *rbdVolume)
-	}
-
-	err = tempClone.checkSnapExists(snap)
+	err := tempClone.checkSnapExists(snap)
```
the commit `rbd: fix clone problem` does not contain any description. It would be good to describe what problem is fixed (and how).

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -19,6 +19,7 @@ package rbd
 import (
 	"context"
 	"errors"
+	"fmt"
```
several typos in the description of commit `rbd: fix snapshot id idempotency issue`

nixpanic reviewed (changes requested):
no issues that I can spot, except for the commit messages that need correcting

Rakshith-R commented:
Removed commit to `set hard and soft limit to lower values in helm`, as this may trigger https://github.com/ceph/ceph-csi/issues/2327 and cause instability in our ci. 

@nixpanic @humblec @Madhu-1 PTAL

Rakshith-R commented:
/restest ci/centos/mini-e2e-helm

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm

`Aug  3 09:47:08.230: INFO: At 2021-08-03 09:37:17 +0100 BST - event for pod-run-as-non-root: {kubelet minikube} Failed: Failed to pull image "registry.centos.org/centos:latest": rpc error: code = Unknown desc = Error response from daemon: Get https://registry.centos.org/v2/: x509: certificate has expired or is not yet valid`

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.19/detail/mini-e2e-helm_k8s-1.19/2287/pipeline/
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/1880/pipeline/

nixpanic commented on e2e/rbd.go:
```diff
@@ -1691,111 +1691,144 @@ var _ = Describe("RBD", func() {
+						f,
+						defaultSCName,
+						nil,
+						map[string]string{
```
There is no test without encryption anymore? If encryption is needed to be tested with this combination, a new test should be added.

Madhu-1 commented:
@Rakshith-R is this PR also fixing https://github.com/ceph/ceph-csi/issues/1883 if yes please add Fixes to the PR description.

Rakshith-R commented:
> @Rakshith-R is this PR also fixing #1883 if yes please add Fixes to the PR description.

Done, PTAL @nixpanic @Madhu-1 @humblec

Rakshith-R commented:
@Mergifyio rebase

Rakshith-R commented:
/retest ci/centos/upgrade-tests-rbd

Rakshith-R commented:
> /retest ci/centos/upgrade-tests-rbd

failed due to #2264  
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/upgrade-tests-rbd/detail/upgrade-tests-rbd/2581/pipeline)

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.20

Rakshith-R commented:
> /retest ci/centos/mini-e2e/k8s-1.20

failed due to #2264
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.20/detail/mini-e2e_k8s-1.20/1996/pipeline/70)

Rakshith-R commented:
@Madhu-1 ,This pr is waiting for your final review/approval PTAL 
Thanks!

Madhu-1 commented on internal/rbd/rbd_journal.go:
```diff
@@ -311,6 +311,7 @@ func (rv *rbdVolume) Exists(ctx context.Context, parentVol *rbdVolume) (bool, er
 
 		return false, err
 	}
+	// TODO: check image needs flattening and completed?
```
create an issue to track this one.

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.21

Rakshith-R commented:
> /retest ci/centos/mini-e2e/k8s-1.21

failed due to #2264 
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.21/detail/mini-e2e_k8s-1.21/671/pipeline)
